### PR TITLE
Migrate string encoding conversions to FML

### PR DESCRIPTION
**Pull request by cbracken**
We've implemented UTF-8/UTF-16 string encoding conversions in multiple
places, from FML to //flutter/shell/platform/common, to the individual
embedders. This migrates these conversions to FML and adds tests.

Windows APIs use wchar_t-based strings and as a result, we continue to
keep Windows-specific functions in fml/platform/win/wstring_conversion.h.

We break out string_conversions into its own source set since FML brings
with it some Dart dependencies (e.g. dart_timestamp_provider.cc) that
are unused by some targets such as uwptool.exe in the Windows UWP
embedding.

Issue: https://github.com/flutter/flutter/issues/98061

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/keyboard_key_embedder_handler.cc [hunk 1]
```diff
@@ -8,12 +8,11 @@
 #include <windows.h>
 
 #include <chrono>
-#include <codecvt>
 #include <iostream>
 #include <string>
 
+#include "flutter/fml/platform/win/wstring_conversion.h"
```
Comment (stuartmorgan) [inline shell/platform/windows/keyboard_key_embedder_handler.cc:14]: Looks like this is unused.

File: fml/string_conversion.cc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// bug is fixed.
+std::locale::id std::codecvt<char16_t, char, _Mbstatet>::id;
+#endif  // defined(FML_OS_WIN)
+
+namespace fml {
+
+using Utf16StringConverter =
+    std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>;
+
+// Returns a UTF-8 encoded equivalent of a UTF-16 encoded input string.
```
Comment (stuartmorgan) [inline fml/string_conversion.cc:25]: Nit: remove the declaration comments here and below; they are in the header.

File: fml/platform/win/wstring_conversion.h [hunk 3]
```diff
@@ -5,24 +5,15 @@
...
 
-#include <codecvt>
-#include <locale>
 #include <string>
 
 namespace fml {
 
-using WideStringConvertor =
-    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>, wchar_t>;
+// Returns a UTF-8 encoded equivalent of a UTF-16 encoded input string.
```
Comment (stuartmorgan) [inline fml/platform/win/wstring_conversion.h:12]: Nit: maybe mention wide string in these comments? They currently read identically to the comments for the non-wide versions which is a bit confusing.

File: fml/platform/win/wstring_conversion.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include <codecvt>
+#include <locale>
+#include <string>
+
+namespace fml {
+
+using WideStringConverter =
+    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>, wchar_t>;
+
+// Returns a UTF-8 encoded equivalent of a UTF-16 encoded input string.
```
Comment (stuartmorgan) [inline fml/platform/win/wstring_conversion.cc:16]: Nit: remove the declaration comments here and below; they are in the header.

File: fml/BUILD.gn [hunk 5]
```diff
@@ -286,6 +319,7 @@ if (enable_unittests) {
       "message_loop_unittests.cc",
       "paths_unittests.cc",
       "raster_thread_merger_unittests.cc",
+      "string_conversion_unittests.cc",
```
Comment (stuartmorgan) [inline fml/BUILD.gn:322]: It looks like you're never building `wstring_conversion_unittests.cc`.

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Thanks for putting in the work to clean this up!

File: fml/platform/win/wstring_conversion.h (hunk 3 above)
Comment (cbracken) [reply fml/platform/win/wstring_conversion.h:12]: Done.

File: fml/string_conversion.cc (hunk 2 above)
Comment (cbracken) [reply fml/string_conversion.cc:25]: Caught in the act of copy-pasting declarations from the headers! Thanks for spotting.

**Pull request merged**